### DISCUSSION TITLE
Formula Column JavaScript with Input from User

### DISCUSSION THREAD

**Discussion by APBowls** [Budibase/budibase#9712] [Help]
Hello everyone,

I'm following up on a thread #9300 where I wrote a JS Function in a formula column to calculate point values. 

![211645118-6764252b-ce17-455a-91dc-8f3d13c16ba7](https://user-images.githubusercontent.com/122388679/219398163-02f9c14b-24ee-498c-9793-d0847b0fbdb5.png)

The Formula in TOTALPOINTS Column:
`(($("Points").map(q => q.Points).reduce((a, b) => a + b))`

![211640591-73701140-5331-4fa4-871a-060a209a89c3](https://user-images.githubusercontent.com/122388679/219399042-066c9cdf-7b6e-46c4-98b6-d13b1bf4d39a.png)

I now need to implement a feature that allows the end user to select a date range, and only perform the calculations on rows within that range. Is there a way to take input from the Production State back into the formula column in the Data Workspace? The goal would be to edit the formula to only iterate through the points values within the date range.

Any help would be greatly appreciated.

**Comment by melohagan** [comment]
Hey @APBowls 

Formula columns only do calculations at the **Data** level. You cannot pass values through to them from the front-end **Design** section.

If you wanted to calculate points for a user specified range then you would need to do that calculation in a JavaScript binding within the **Design** section.

So instead of getting relationship data through a formula, you would get it through a front-end repeater:
https://docs.budibase.com/docs/fetching-data#displaying-relational-data

The filtering on the front-end would be limited by the Data provider pagination, however if the user is filtering the data then hopefully this won't be an issue if you set the limit and pagination sufficiently high enough.

**Reply by APBowls** [reply]
The user filtering would be taking into account two weeks worth of production, anywhere between 5 - 75 rows at a time. Also, I am not able to figure out a front-end repeater. I have the relationship I want, which is the employee with points, hours, and bonus displayed. Yet some of the employees are duplicated more times than others, seemingly randomly.
![Capture](https://user-images.githubusercontent.com/122388679/220124459-4784bdf3-8120-4237-9c89-69466ac6beaa.PNG)

**Reply by melohagan** [reply]
It looks like you have a Table component inside a Repeater component. You probably don't want to do that.

Could you share a screenshot of your component tree please, alongside the data source for each data provider and any filters that are being applied.

**Reply by APBowls** [reply]
![Capture](https://user-images.githubusercontent.com/122388679/220125953-6db6c17f-9508-4ec4-b563-de5f38bd4ac2.PNG)

![Capture 1](https://user-images.githubusercontent.com/122388679/220125896-a2e0e0f7-d843-4b59-be68-c05b241d9968.PNG)

![Bi weekly production data source](https://user-images.githubusercontent.com/122388679/220126509-c5a81d6a-716f-43ee-b146-bb5da0fe9395.PNG)

![pay period table](https://user-images.githubusercontent.com/122388679/220126512-3500d2ff-6400-43f3-9ce6-13d3e542acd7.PNG)

**Reply by melohagan** [reply]
So I think to do what you're trying to do would be to allows the user to enter a data range into your form, and these date fields would be the basis of your filter for the **Data provider** (for the Bi-Weekly Production table). 

Next I would delete the **Repeater** and **Table block**. Instead add a container with some text elements (or a card component).

Then you can use a JavaScript binding to calculate the total points from the Data provider data. 
 ```js
(($("Dataprovider Rows").map(q => q.Points).reduce((a, b) => a + b))
```

**Reply by APBowls** [reply]
Yes! That is exactly what I'm looking to do. I'll give that a try. Thank you!

**Reply by APBowls** [reply]
I'm running into a JavaScript error. Is there possibly something wrong with the binding? I tried 
`(($("Dataprovider Rows").map(q => q.Points).reduce((a, b) => a + b))`
`(($("Dataprovider.Rows").map(q => q.Points).reduce((a, b) => a + b))`

![error js](https://user-images.githubusercontent.com/122388679/220132442-235a8c7b-884f-4681-b424-2862187132c8.PNG)

**Reply by melohagan** [reply]
Did you try adding the Data provider rows binding from the bindings helper in the left-hand side panel within the filter drawer (rather than manually type it out.

Is `Points` definitely the case-sensitive correct name for the column in the Bi-Weekly Production table. Is the Dataprovider definitely using this table?

**Reply by APBowls** [reply]
Oh, I think I see the problem. I am doing additional math in the JavaScript Binding which references a different table. Any way to relate both tables to the one card component?

**Reply by melohagan** [reply]
Hmm, would you be able to provide me with screenshots of the different tables involved and the calculation you want to do between the tables please?

Then I can try and replicate it myself.

**Reply by APBowls** [reply]
I can do an app export if that is any help. It is all a bit messy, I'm still getting the hang of this.

**Reply by melohagan** [reply]
Yeah the app export would be great - you can send it to support@budibase.com

**Comment by melohagan** [comment]
Hey @APBowls 

Thank you for the app export. I believe I've figured this one out, although I'm not certain that it's the use-case you're after.

It was a little bit tricky. 

First I ejected the **Card Block** you had added, and nested another Data Provider for the *Employees* relationship. This was to allow me access to the *Regular Hours* and *Research Hours*.

![Screenshot 2023-02-20 at 15 28 33](https://user-images.githubusercontent.com/101575380/220146674-7dd9c9f7-2af4-41ee-9182-c2b7c8e00ed0.png)
![Screenshot 2023-02-20 at 15 28 58](https://user-images.githubusercontent.com/101575380/220146742-f38a77b7-d45b-4c0a-a8b8-a75a5ea887ec.png)

Then the JavaScript binding for the card description looks as follows:

```js
let regularHours = $("Employees Data Provider.Rows").find(emp => emp._id === $("Repeater.Bi-Weekly Production.Employees")[0]._id)["Regular Hours"]

//pph stands for 'Points Per Hour', Regular Hours reference from Pay Period Table
let pph = ($("Dataprovider.Rows").map(q => q.Points).reduce((a, b) => a + b)) / regularHours.toFixed(1)

//18.00 is pay for research and development hours, referenced from Pay Period Table
let rd = 18 * $("Employees Data Provider.Rows").find(emp => emp._id === $("Repeater.Bi-Weekly Production.Employees")[0]._id)["Research and Development Hours"]

if (pph < 8) {
  return (0.00 * regularHours) + rd

...etc
```

**Reply by melohagan** [reply]
To be honest this particular use-case would be easier with SQL, because then you could do the JOINs you need with a [Custom query](https://docs.budibase.com/docs/custom-queries)

**Reply by APBowls** [reply]
I have not been able to get this to work unfortunately. I will look into SQL, I'm not particularly knowledgeable about it, hence attempting to use internal tables. Thank you for taking a stab at it, I'm sure it's something small I'm not seeing.
